test/test_lib.py

```diff
@@ -0,0 +1,442 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+
+from collections import namedtuple
+from importlib.machinery import ModuleSpec
+from pathlib import Path
+from tempfile import mkdtemp, mkstemp
+from unittest import TestCase
+from unittest.mock import patch
+from random import shuffle
+from shutil import rmtree
+from textwrap import dedent
+
+import ops.lib
+
+
+def _mklib(topdir: str, pkgname: str, libname: str) -> Path:
+    """Make a for-testing library.
+
+    Args:
+        topdir: the toplevel directory in which the package will be created.
+             This directory must already exist.
+        pkgname: the name of the package to create in the toplevel directory.
+             this package will have an empty __init__.py.
+        libname: the name of the library directory to create under the package.
+
+    Returns:
+        a :class:`Path` to the ``__init__.py`` of the created library.
+        This file will not have been created yet.
+    """
+    pkg = Path(topdir) / pkgname
+    try:
+        pkg.mkdir()
+    except FileExistsError:
+        pass
+    else:
+        (pkg / '__init__.py').write_text('')
+
+    lib = pkg / 'opslib' / libname
+    lib.mkdir(parents=True)
+
+    return lib / '__init__.py'
+
+
+def _flatten(specgen):
+    return [os.path.dirname(spec.origin) for spec in specgen]
+
+
+class TestLibFinder(TestCase):
+    def _mkdtemp(self) -> str:
+        tmpdir = mkdtemp()
+        self.addCleanup(rmtree, tmpdir)
+        return tmpdir
+
+    def test_single(self):
+        tmpdir = self._mkdtemp()
+
+        self.assertEqual(list(ops.lib._find_all_specs([tmpdir])), [])
+
+        _mklib(tmpdir, "foo", "bar").write_text("")
+
+        self.assertEqual(
+            _flatten(ops.lib._find_all_specs([tmpdir])),
+            [tmpdir + '/foo/opslib/bar'])
+
+    def test_multi(self):
+        tmpdirA = self._mkdtemp()
+        tmpdirB = self._mkdtemp()
+
+        dirs = [tmpdirA, tmpdirB]
+
+        expected = []
+        for top in [tmpdirA, tmpdirB]:
+            for pkg in ["bar", "baz"]:
+                for lib in ["quux", "meep"]:
+                    _mklib(top, pkg, lib).write_text("")
+                    expected.append(os.path.join(top, pkg, "opslib", lib))
+
+        self.assertEqual(
+            sorted(_flatten(ops.lib._find_all_specs(dirs))),
+            sorted(expected))
```

Comment:
yep. Would it be cleaner if I wrote that out by hand?

Reply:
pros/cons

a) It feels a bit wrong to repeat yourself
b) It is easier to maintain the test with the way you wrote it
c) It makes it really easy to accidentally assertEqual([], []) because you missed that your for loop was exiting for some reason
d) 'tmpdir' being a random prefix on each iteration means you don't just have clean strings to compare. (At the same time, that means if it ever fails, you end up with a lot of random chars in the failure output.)
e) I do like writing tests like this and making sure they fail, so that I know what the failure output looks like and whether it is understandable when you want to come back and debug it later.

IMO this falls into the "enough pro/con that the one doing the implementation gets to decide", but I wanted to share some thoughts on it.